src/specs/FrameManager.spec.ts

```diff
@@ -0,0 +1,214 @@
+      expect(mocks.frame.contentWindow.postMessage).toHaveBeenCalledWith(
+        message,
+        'http://example.com:4040'
+      );
+    });
+
+    it('or restricted by the implied host page origin', () => {
+      const message = {
+        msgType: 'publish',
+        msg: {
+          topic: 'test.topic',
+          payload: {}
+        }
+      } as HostToClient;
+      frameManager.setFrameLocation('/foo/bar/');
+      frameManager.sendToClient(message);
+
+      expect(mocks.frame.contentWindow.postMessage).toHaveBeenCalledWith(
+        message,
+        'http://test.example.com'
+      );
+    });
+  });
+
+  describe('Can be used to subscribe to client messages', () => {
+    beforeEach(() => {
+      mocks.handler = jasmine.createSpy('messageHandler');
+      mocks.frame.load();
+      frameManager.setFrameLocation('http://example.com/');
+      mocks.messageEvent = {
+        origin: 'http://example.com',
+        source: mocks.frame.contentWindow,
+        data: {
+          msgType: 'publish',
+          msg: {
+            topic: 'test.topic',
+            payload: {}
+          }
+        }
+      };
+      frameManager.listenToMessages(mocks.handler);
+    });
+
+    it('if the messages are sent correctly', () => {
+      mocks.window.raise('message', mocks.messageEvent);
+      expect(mocks.handler).toHaveBeenCalledWith(mocks.messageEvent.data);
+    });
+
+    it('which are blocked if sent from an unexpected origin', () => {
+      mocks.messageEvent.origin = 'http://evil.com';
+      mocks.window.raise('message', mocks.messageEvent);
+      expect(mocks.handler).not.toHaveBeenCalled();
+    });
+
+    it('which are blocked if sent from an unexpected frame', () => {
+      mocks.messageEvent.source = {};
+      mocks.window.raise('message', mocks.messageEvent);
+      expect(mocks.handler).not.toHaveBeenCalled();
+    });
+
+    it('which are blocked if sent with invalid data', () => {
+      mocks.messageEvent.data = { msgType: 'notValid', msg: {} };
+      mocks.window.raise('message', mocks.messageEvent);
+      expect(mocks.handler).not.toHaveBeenCalled();
+    });
+  });
+
+  describe('Can send messages to clients', () => {
+    beforeEach(() => {
+      mocks.frame.load();
+      frameManager.setFrameLocation('http://example.com/');
+    });
+
+    it('if the message is valid', () => {
+      const message = {
+        msgType: 'publish',
+        msg: {
+          topic: 'test.topic',
+          payload: {}
+        }
+      } as HostToClient;
+
+      frameManager.sendToClient(message);
+      expect(mocks.frame.contentWindow.postMessage).toHaveBeenCalledWith(
+        message,
+        'http://example.com'
+      );
+    });
+
+    /*
+    it('but not if the message is invalid', () => {
+      const message = {
+        msgType: 'notAValidType',
+        msg: {}
+      } as HostToClient;
+      frameManager.sendToClient(message);
+      expect(mocks.frame.contentWindow.postMessage).not.toHaveBeenCalled();
+    });
```

Comment:
I would like to test this, since it's a use case when called from JavaScript, but I can't get TypeScript to let me pass invalid data in the test!  Any suggestions?